judge/views/contests.py

```diff
@@ -76,17 +76,22 @@ def get_queryset(self):
     def get_context_data(self, **kwargs):
         context = super(ContestList, self).get_context_data(**kwargs)
         now = timezone.now()
-        past, present, future = [], [], []
+        past, present, active, future = [], set(), [], []
         for contest in self.get_queryset():
             if contest.end_time < now:
                 past.append(contest)
             elif contest.start_time > now:
                 future.append(contest)
             else:
-                present.append(contest)
+                present.add(contest)
+        if self.request.user.is_authenticated:
+            active = filter(lambda x: not x.ended, ContestParticipation.objects.filter(contest__in=present, virtual=0))
+            present -= set(map(lambda x: x.contest, active))
```

Comment:
This runs one query per active contest to fetch a full contest object, as well as a query to fetch the participation that you are ignoring. You should just fetch the contest ids and discard contests based on the ids.